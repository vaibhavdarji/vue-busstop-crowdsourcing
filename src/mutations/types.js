export const SET_STOPS = 'SET_STOPS';
export const SET_IS_LOADING = 'SET_IS_LOADING';
export const SET_SELECTED_STOP = 'SET_SELECTED_STOP';
export const SET_TOTAL_FUNDS = 'SET_TOTAL_FUNDS';
export const UPDATE_USER_DONATION_INFO = 'UPDATE_USER_DONATION_INFO';
export const UPDATE_BUS_STOP_DONATION = 'UPDATE_BUS_STOP_DONATION';
export const CLEAR_SELECTED_BUS_STOP = 'CLEAR_SELECTED_BUS_STOP';
