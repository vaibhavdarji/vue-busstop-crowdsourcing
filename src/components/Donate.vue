<template>
  <v-card>
    <v-form ref="form" @submit.prevent="onNext">
      <v-container grid-list-xl fluid>
        <v-layout wrap>
           <v-flex xs12>
             <v-card-title class="headline pl-0">
               Yes, I want to donate to
              {{selectedBusStop.name}} :
              <template>
                <v-icon left>attach_money</v-icon>
                {{form.amount}}
              </template>
            </v-card-title>
           </v-flex>
           <v-flex xs12 pl-0>
             <template v-for="(option,index) in Constants.DONATE_OPTIONS">
               <custom-button
                :key="index"
                @onClick="onSelectDonateOption(option)">
                 <v-icon left>attach_money</v-icon>
                {{option}}
             </custom-button>
              </template>
           </v-flex>
           <v-flex xs12 sm6>
             <v-text-field 
              label="Enter Your Amount" 
              prepend-inner-icon="attach_money" 
              class="inputPrice"
              outline
              type="number"
              ref="otherAmount"
              @input="OnInputOtherAmount"
            ></v-text-field>
           </v-flex>
           <v-flex xs12 v-if="!formIsValid">
             <span :class="Constants.THEME.ERROR_TEXT_CLASS">Please select an amount to donate</span>
           </v-flex>
           <v-flex xs12 pl-0>
             <custom-button type="submit" :disabled="!formIsValid">
               Next 
             </custom-button>
           </v-flex>
        </v-layout>
      </v-container>
    </v-form>
  </v-card>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';
import CustomButton from '@/components/Button.vue';
import Constants from '@/constants';
import {
  UPDATE_USER_DONATION_DETAIL
} from '@/actions/types';

export default {
  name: 'Donate',
  data () {
    const defaultForm = Object.freeze({
      amount: 0
    });
    return {
      form: Object.assign({}, defaultForm),
      rules: {
        amount: [ val => val >= 1]
      },
      defaultForm
    };
  },

  components: {
    CustomButton
  },

  computed: {
    ...mapGetters(['selectedBusStop']),
    Constants() {
      return Constants;
    },
    formIsValid () {
      return (
        this.form.amount >= 1
      )
    }
  },

  methods: {
    ...mapActions({
      'updateUserDonationDetail': UPDATE_USER_DONATION_DETAIL
    }),
    onSelectDonateOption (value) {
      this.$refs.otherAmount.reset();
      this.form.amount = value;
    },

    OnInputOtherAmount (value) {
      this.form.amount = parseFloat(parseFloat(value).toFixed(2));
    },

    onNext () {
      this.$emit('onNextStep');
      this.updateUserDonationDetail({amount: this.form.amount});
    },
  }
}
</script>

<style >
  .inputPrice 
    input[type='number'] {
      -moz-appearance: textfield;
  }

  .inputPrice
     input::-webkit-outer-spin-button,
     input::-webkit-inner-spin-button {
      -webkit-appearance: none;
  }
</style>


