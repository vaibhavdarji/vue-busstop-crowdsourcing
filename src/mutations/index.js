import * as MUTATION_TYPE from '@/mutations/types';

export default {
  [MUTATION_TYPE.SET_STOPS] (state, payload) {
    state.busStops = payload;
  },

  [MUTATION_TYPE.SET_IS_LOADING] (state, payload) {
    state.isLoading = payload;
  },

  [MUTATION_TYPE.SET_SELECTED_STOP] (state, stop) {
    state.selectedBusStop = stop;
  },

  [MUTATION_TYPE.SET_TOTAL_FUNDS] (state, payload) {
    state.totalFunds += payload;
  },

  [MUTATION_TYPE.UPDATE_USER_DONATION_INFO] (state, payload) {
    Object.assign(state.userDonation, payload);
  },

  [MUTATION_TYPE.UPDATE_BUS_STOP_DONATION] (state) {
    Object.assign(state.selectedBusStop, {
      donationsRaisedInDollars: state.selectedBusStop.donationsRaisedInDollars + state.userDonation.amount
    });
    Object.assign(state.busStops, state.busStops.map(el=> el.stopId === state.selectedBusStop.stopId? state.selectedBusStop : el))
  },

  [MUTATION_TYPE.CLEAR_SELECTED_BUS_STOP] (state) {
    state.selectedBusStop = null;
  }
};