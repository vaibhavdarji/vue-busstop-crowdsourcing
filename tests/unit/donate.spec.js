import { shallowMount, mount } from "@vue/test-utils";
import Donate from '@/components/Donate.vue';
import storeConfig from '@/store';
import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";
import {
  SET_SELECTED_STOP
} from '@/actions/types';
import {
  SET_STOPS,
  UPDATE_BUS_STOP_DONATION
} from '@/mutations/types';
import mutations from '@/mutations';

import Constants from '@/constants';

Vue.use(Vuex);
Vue.use(Vuetify);

const dummyData = [
  {stopId: 1, lat: 33.760262, lng: -84.384706, donationsRaisedInDollars: 0, name: 'Hertz at Portman Blvd'},
  { stopId: 2, lat: 33.760138, lng: -84.388043, donationsRaisedInDollars: 0, name: 'Peachtree Center Mall' }
];

const currentStop  = dummyData[1];

describe("Donate.vue", () => {
  let store;
  beforeEach(() => {
    store = Object.assign({}, storeConfig);
  });
  it("should match snapshot", () => {
    store.commit(SET_STOPS, dummyData);
    store.dispatch(SET_SELECTED_STOP, currentStop);
    const wrapper = shallowMount(Donate, {
      store,
      Vue
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("validate selected bus stop", () => {
    const wrapper = shallowMount(Donate, {
      store,
      Vue
    });
    expect(wrapper.vm.selectedBusStop).toEqual(store.getters.selectedBusStop);
  });

  it("should contain donate button list options", () => {
    const wrapper = mount(Donate, {
      store,
      Vue
    });

    expect(wrapper.findAll('button[type="button"]').exists()).toBe(true);
    expect(wrapper.findAll('button[type="button"]').length).toEqual(Constants.DONATE_OPTIONS.length);
  });

  it("should validate data amount is not set", () => {
    const wrapper = mount(Donate, {
      store,
      Vue
    });

    const form = {amount: 0};

    wrapper.setData({
      form
    });

    expect(wrapper.find('.headline').text()).toContain(form.amount);
  });

  it("should trigger any selected donation button and validate data form amount", () => {
    
    
    const wrapper = mount(Donate, {
      store,
      Vue
    });

    const form = {amount: 0};

    wrapper.setData({
      form
    });
    const donateBtn = Constants.DONATE_OPTIONS[Math.round(Math.random()*(Constants.DONATE_OPTIONS.length - 1))];
    const donateBtnIndex = Constants.DONATE_OPTIONS.indexOf(donateBtn);
    const wrapperBtn = wrapper.findAll('button[type="button"]').at(donateBtnIndex);
    const donateInput = wrapper.find('input[type="number"]');
    // user added input number value but selecting button option.
    // so validating input value reset and button amount is selected and form amount match
    
    donateInput.setValue('20');
    expect(donateInput.element.value).toBe('20');
    wrapperBtn.trigger('click');
    expect(wrapper.vm.form.amount).toEqual(donateBtn);
    expect(donateInput.element.value).toBe('');
  });

  it("should validate form submit button disable prop", () => {
    const wrapper = mount(Donate, {
      store,
      Vue
    });
    const form = {amount: 0};
    wrapper.setData({
      form
    });

    const donateInput = wrapper.find('input[type="number"]');
    const nextBtn = wrapper.find('button[type="submit"]');

    // form submit button disabled reason no amount selected
    expect(nextBtn.is('[disabled]')).toBe(true);
    donateInput.setValue('30');
    expect(donateInput.element.value).toBe('30');
    // amount is set, form buttom should not be disabled
    expect(nextBtn.is('[disabled]')).toBe(false);
  });

  it("should trigger form submit click once amount is set", () => {
    const wrapper = mount(Donate, {
      store,
      Vue
    });
    const form = {amount: 0};
    wrapper.setData({
      form
    });

    const donationAmount = 50;

    const donateInput = wrapper.find('input[type="number"]');
    const donateForm = wrapper.find('form');

    donateInput.setValue(donationAmount.toString());
    expect(donateInput.element.value).toBe(donationAmount.toString());

    donateForm.trigger('submit');
    store.commit(UPDATE_BUS_STOP_DONATION, donationAmount);

    // validate user donation amount has changed
    expect(store.getters.userDonation.amount).toBe(donationAmount);
    expect(store.getters.totalFunds).toBe(donationAmount);
    
  });
});