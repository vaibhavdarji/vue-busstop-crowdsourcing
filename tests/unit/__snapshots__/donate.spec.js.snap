// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Donate.vue should match snapshot 1`] = `
<v-card-stub
  tag="div"
>
  <v-form-stub>
    <v-container-stub
      fluid=""
      grid-list-xl=""
      tag="div"
    >
      <v-layout-stub
        tag="div"
        wrap=""
      >
        <v-flex-stub
          tag="div"
          xs12=""
        >
          <v-card-title-stub
            class="headline pl-0"
          >
            
             Yes, I want to donate to
            Peachtree Center Mall :
            
            <v-icon-stub
              left=""
            >
              attach_money
            </v-icon-stub>
            
              0
            
          </v-card-title-stub>
        </v-flex-stub>
         
        <v-flex-stub
          pl-0=""
          tag="div"
          xs12=""
        >
          <custom-button-stub
            btntype="button"
          >
            <v-icon-stub
              left=""
            >
              attach_money
            </v-icon-stub>
            
              2
           
          </custom-button-stub>
          <custom-button-stub
            btntype="button"
          >
            <v-icon-stub
              left=""
            >
              attach_money
            </v-icon-stub>
            
              5
           
          </custom-button-stub>
          <custom-button-stub
            btntype="button"
          >
            <v-icon-stub
              left=""
            >
              attach_money
            </v-icon-stub>
            
              10
           
          </custom-button-stub>
          <custom-button-stub
            btntype="button"
          >
            <v-icon-stub
              left=""
            >
              attach_money
            </v-icon-stub>
            
              15
           
          </custom-button-stub>
          <custom-button-stub
            btntype="button"
          >
            <v-icon-stub
              left=""
            >
              attach_money
            </v-icon-stub>
            
              20
           
          </custom-button-stub>
          <custom-button-stub
            btntype="button"
          >
            <v-icon-stub
              left=""
            >
              attach_money
            </v-icon-stub>
            
              50
           
          </custom-button-stub>
        </v-flex-stub>
         
        <v-flex-stub
          sm6=""
          tag="div"
          xs12=""
        >
          <v-text-field-stub
            class="inputPrice"
            label="Enter Your Amount"
            outline=""
            prepend-inner-icon="attach_money"
            type="number"
          />
        </v-flex-stub>
         
        <v-flex-stub
          tag="div"
          xs12=""
        >
          <span
            class="error--text title font-weight-bold"
          >
            Please select an amount to donate
          </span>
        </v-flex-stub>
         
        <v-flex-stub
          pl-0=""
          tag="div"
          xs12=""
        >
          <custom-button-stub
            btntype="button"
            disabled="true"
            type="submit"
          >
            
             Next 
           
          </custom-button-stub>
        </v-flex-stub>
      </v-layout-stub>
    </v-container-stub>
  </v-form-stub>
</v-card-stub>
`;
